src/compiler/parser.ts

```diff
@@ -7044,6 +7042,11 @@ namespace ts {
                         }
                     }
 
+                    if (constraint) {
+                        Debug.assert(!!typeParameters.length);

```

Comment:
`first(typeParameters)` will do this assertion for you.

Reply:
Thanks, didn't think about using `first(typeParameters).constraint` as an lhs